packages/ampd-sdk/src/event/event_handler.rs

```diff
@@ -0,0 +1,520 @@
+use std::fmt::Display;
+use std::time::Duration;
+
+use async_trait::async_trait;
+use cosmrs::Any;
+use error_stack::{report, Context, Report, ResultExt};
+use events::{AbciEventTypeFilter, Event};
+use futures::{pin_mut, Stream, TryFutureExt};
+use mockall::automock;
+use report::ErrorExt;
+use serde::{Deserialize, Serialize};
+use thiserror::Error;
+use tokio::time::interval;
+use tokio_stream::{Elapsed, StreamExt};
+use tokio_util::sync::CancellationToken;
+use tracing::{error, info};
+use typed_builder::TypedBuilder;
+
+use crate::event::callbacks::{
+    default_broadcast_error_cb, default_event_subscription_error_cb, default_handler_error_cb,
+};
+use crate::future::{with_retry, RetryPolicy};
+use crate::grpc::client;
+
+#[automock(
+    type Err = Error;
+    type Event = Event;
+)]
+#[async_trait]
+pub trait EventHandler: Send + Sync {
+    type Err: Context;
+    type Event: TryFrom<Event>;
+
+    async fn handle(
+        &self,
+        event: &Self::Event,
+        token: &CancellationToken,
+    ) -> Result<Vec<Any>, Self::Err>;
+
+    fn subscription_params(&self) -> SubscriptionParams;
+}
+
+pub struct SubscriptionParams {
+    event_filters: Vec<AbciEventTypeFilter>,
+    include_block_begin_end: bool,
+}
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct Config {
+    #[serde(with = "humantime_serde")]
+    pub stream_timeout: Duration,
+}
+
+impl Default for Config {
+    fn default() -> Self {
+        Self {
+            stream_timeout: Duration::from_secs(10),
+        }
+    }
+}
+
+#[derive(Error, Debug)]
+#[allow(dead_code)]
+pub enum Error {
+    #[error("failed to retrieve events stream from the client")]
+    EventStream,
+
+    #[error("timeout while waiting for event stream")]
+    StreamTimeout(#[from] Elapsed),
+
+    #[error("unable to parse event of type")]
+    EventConversion,
+
+    #[error("error when handling event messages")]
+    HandlerFailed,
+
+    #[error("error when broadcasting message")]
+    BroadcastFailed,
+
+    #[error("{0} task aborted")]
+    TaskAborted(String),
+}
+
+#[allow(dead_code)]
+pub enum HandlerTaskAction {
+    Abort,
```

Comment:
I have had this thought of 
```suggestion
    Abort(Error),
```
so that you can return the actual error in the end instead of using something so generic like `TaskAborted`.